lib/rules/require-presentational-children.js

```diff
@@ -0,0 +1,101 @@
+'use strict';
+
+const AstNodeInfo = require('../helpers/ast-node-info');
+const { flatMap } = require('../helpers/javascript');
+const Rule = require('./base');
+
+function createErrorMessage(ancestorElement, role, descendantElement) {
+  return `<${ancestorElement.tag}> has a role of ${role}, it cannot have semantic descendants like <${descendantElement.tag}>`;
+}
+
+/**
+ * Gets all the element node descendants. If it is presentational we do not include it in the list
+ * @param {object} node
+ * @returns Array of element nodes
+ */
+function getAllElementNodeDescendants(node) {
+  return flatMap(AstNodeInfo.childrenFor(node), (childNode) => {
+    const isNotPresentationRole =
+      AstNodeInfo.attributeTextValue(AstNodeInfo.findAttribute(childNode, 'role')) !==
+      'presentation';
+    let descendants = [];
+
+    if (AstNodeInfo.hasChildren(childNode)) {
+      descendants = getAllElementNodeDescendants(childNode);
+    }
+
+    if (isNotPresentationRole && childNode.type === 'ElementNode') {
+      return [childNode, ...descendants];
+    }
+
+    return [...descendants];
+  });
+}
+
+// List of roles that cannot support semantic children
+// https://w3c.github.io/aria-practices/#children_presentational
+const ROLES_THAT_CANNOT_SUPPORT_SEMANTIC_CHILDREN = new Set([
+  'button',
+  'checkbox',
+  'img',
+  'meter',
+  'menuitemcheckbox',
+  'menuitemradio',
+  'option',
+  'progressbar',
+  'radio',
+  'scrollbar',
+  'separator',
+  'slider',
+  'switch',
+  'tab',
+]);
+
+const NON_SEMANTIC_TAGS = new Set([
```

Comment:
We could do it this way, but I wonder if we should make a helper that has all of the semantic elements in it, and a helper that has all valid roles for author use it it. That way we can standardize across the app. 

cc'ing @Turbo87  and @rwjblue for input here.

Reply:
Right like a central const file. 

I chose to check if the tag is non-semantic vs semantic because there are a lot of semantic tags and the list can grow.